loki/transformations/temporaries/stack_allocator.py

```diff
@@ -0,0 +1,1372 @@
+            The items corresponding to successor routines called from :data:`routine`
+        """
+
+        stack_vars = []
+        stack_args = []
+        stack_arg_dict = {}
+        pragma_string = ''
+        assignments = []
+        for dtype in stack_dict:
+            for kind in stack_dict[dtype]:
+
+                # start arguments integer names in kernels with 'K'
+                stack_size_name = self._get_stack_int_name('K', dtype, kind, 'STACK_SIZE')
+                stack_size_var = self._get_int_var(name=stack_size_name, scope=routine,
+                                                   type=self._get_int_type(intent='IN'))
+
+                # local variables start with 'J'
+                stack_used_arg_name = self._get_stack_int_name('JD', dtype, kind, 'STACK_USED')
+                stack_used_arg = self._get_int_var(name=stack_used_arg_name, scope=routine,
+                                                   type=self._get_int_type(intent='INOUT'))
+                stack_used_name = self._get_stack_int_name('J', dtype, kind, 'STACK_USED')
+                stack_used_var = self._get_int_var(name=stack_used_name, scope=routine)
+                assignments += [Assignment(lhs=stack_used_var, rhs=stack_used_arg)]
+
+                # create the stack variable and its type with the correct shape
+                shape = (stack_size_var,)
+                stack_var = self._get_stack_var(routine, dtype, kind)
+                stack_type = stack_var.type.clone(shape=as_tuple(shape), target=True)
+                stack_var = stack_var.clone(type=stack_type)
+
+                # pass on the stack variable from stack_used + 1 to stack_size
+                # pass stack_size - stack_used to stack size in called kernel
+                arg_dims = (RangeIndex((None, None)),)
+                stack_arg_dict.setdefault(dtype, {})
+                stack_arg_dict[dtype][kind] = (stack_size_var, stack_var.clone(dimensions=arg_dims), stack_used_var)
+
+                # create stack_vars pair
+                stack_args += [stack_size_var,
+                               stack_var.clone(dimensions=stack_type.shape, type=stack_var.type.clone(contiguous=True)),
+                               stack_used_arg]
+                stack_vars += [stack_used_var]
+                pragma_string += f'{stack_var.name}, '
+
+        if pragma_string:
+            # remove ', '
+            pragma_string = pragma_string[:-2].lower()
+
+            present_pragma = None
+            acc_pragmas = [p for p in FindNodes(Pragma).visit(routine.body) if p.keyword.lower() == 'loki'] # acc
+            for pragma in acc_pragmas:
+                if pragma.content.lower().startswith('device-present'):
+                    present_pragma = pragma
+                    break
+            if present_pragma:
+                pragma_map = {present_pragma: None}
+                routine.body = Transformer(pragma_map).visit(routine.body)
```

Comment:
Not entirely sure I'm following what's happening here. My understanding is, we're
1. Looking for existing `!$loki device-present` pragmas
2. Remove the first such pragma and re-insert it at the beginning of the body with the var list amended
3. We don't add a corresponding `end` pragma
4. If no such pragma exists, we insert a `device-present` pragma pair around the body

Apologies if these are stupid questions:
1. are we certain there's always just one such pragma? 
2. If there can only be one, is there any reason to expect this pragma to be in the wrong place? Because then we could simply do an in-place `_update` with the amended string here, thus speeding up processing and removing some control flow complexity

Reply:
Not stupid at all. Thinking about it we can probably just add the pragmas around the body either way, since having multiple of those device-present pragmas shouldn't be a problem at all and it shouldn't happen that one of the stack variables ends up twice in the pragmas ... What do you think?